src/compiler/parser.ts

```diff
@@ -1284,6 +1318,8 @@ module ts {
                     return token === SyntaxKind.GreaterThanToken || token === SyntaxKind.OpenParenToken;
                 case ParsingContext.HeritageClauses:
                     return token === SyntaxKind.OpenBraceToken || token === SyntaxKind.CloseBraceToken;
+                case ParsingContext.JsxAttributes:
+                    return token === SyntaxKind.GreaterThanToken || token === SyntaxKind.SlashToken || token === SyntaxKind.CloseBraceToken;

```

Comment:
I think we want this for error recovery reasons -- otherwise an over-braced {expr}} will close the surrounding statement block, which is going to mess up the entire file.